src/core/lwt_seq.mli

```diff
@@ -0,0 +1,86 @@
+  not a [lazy] block, which means it might be recomputed every time
+  we access it. *)
+
+and +'a node = Nil | Cons of 'a * 'a t
+(** A fully-evaluated list node, either empty or containing an element
+  and a delayed tail. *)
+
+val empty : 'a t
+(** The empty sequence, containing no elements. *)
+
+val return : 'a -> 'a t
+(** The singleton sequence containing only the given element. *)
+
+val cons : 'a -> 'a t -> 'a t
+(** [cons x xs] is the sequence containing the element [x] followed by
+  the sequence [xs] *)
+
+val append : 'a t -> 'a t -> 'a t
+(** [append xs ys] is the sequence [xs] followed by the sequence [ys] *)
+
+val map : ('a -> 'b Lwt.t) -> 'a t -> 'b t
+(** [map f seq] returns a new sequence whose elements are the elements of
+  [seq], transformed by [f].
+  This transformation is lazy, it only applies when the result is traversed. *)
+
+val filter : ('a -> bool Lwt.t) -> 'a t -> 'a t
+(** Remove from the sequence the elements that do not satisfy the
+  given predicate.
+  This transformation is lazy, it only applies when the result is
+  traversed. *)
+
+val filter_map : ('a -> 'b option Lwt.t) -> 'a t -> 'b t
+(** Apply the function to every element; if [f x = None] then [x] is dropped;
+  if [f x = Some y] then [y] is returned.
+  This transformation is lazy, it only applies when the result is
+  traversed. *)
+
+val flat_map : ('a -> 'b t Lwt.t) -> 'a t -> 'b t
+(** Map each element to a subsequence, then return each element of this
+  sub-sequence in turn.
+  This transformation is lazy, it only applies when the result is
+  traversed. *)
+
+val fold_left : ('a -> 'b -> 'a Lwt.t) -> 'a -> 'b t -> 'a Lwt.t
+(** Traverse the sequence from left to right, combining each element with the
+  accumulator using the given function.
+  The traversal happens immediately and will not terminate on infinite
+  sequences. *)
+
+
+val iter : ('a -> unit Lwt.t) -> 'a t -> unit Lwt.t
+(** Iterate on the sequence, calling the (imperative) function on every element.
+  The traversal happens immediately and will not terminate on infinite
+  sequences. *)
+
+val unfold : ('b -> ('a * 'b) option Lwt.t) -> 'b -> 'a t
+(** Build a sequence from a step function and an initial value.
+  [unfold f u] returns [empty] if the promise [f u] resolves to [None],
+  or [fun () -> Lwt.return (Cons (x, unfold f y))] if the promise [f u] resolves
+  to [Some (x, y)]. *)
+
+val to_list : 'a t -> 'a list Lwt.t
+(** Convert a sequence to a list, preserving order.
+  The traversal happens immediately and will not terminate on infinite
+  sequences. *)
+
+val of_list : 'a list -> 'a t
+(** Convert a list to a sequence, preserving order. *)
+
+val to_seq : 'a t -> 'a Seq.t Lwt.t
+(** Convert from ['a Lwt_seq.t] to ['a Stdlib.Seq.t] *)
+
+val of_seq : 'a Seq.t -> 'a t
+(** Convert from ['a Stdlib.Seq.t] to ['a Lwt_seq.t].
+  This transformation is lazy, it only applies when the result is
+  traversed. *)
```

Comment:
I'm not sure which `of_seq` we should have: the one currently in the MR, another, both?

Options:
- current code
    - `val of_seq : 'a Seq.t -> 'a t`
- Lwt-flipping
    - `val of_seq : 'a Lwt.t Seq.t -> 'a t`

The idea behind the Lwt-flipping `of_seq` is as follows.

You can already have lwt-sequence with `'a Lwt.t Seq.t`. And as long as you are careful, you can consume the promises sequentially:

```
let rec iter_s f s =
  match s () with
  | Nil -> Lwt.return_unit
  | Cons (p, s) ->
    let* x = p in
    let* () = f x in
    iter_s f s
```

This works but you may accidentally step on your own feet and accidentally forget to wait for one promise to resolve before you start evaluating the next one.

The Lwt-flipping `of_seq` allows you to go from the valid-but-error-prone form into the Lwt-specialised version.

Maybe we need both because this Lwt-flipping `of_seq` is not as much as an inverse to `to_seq`. Maybe we can have `of_seq` as is, and we can have `inject` (or some better name) for the Lwt-flipping `of_seq`.

Ideas? Comments? Suggestions?

Reply:
The current `of_seq` implementation should probably stay as it is, but adding the Lwt-flipping version is a good idea. 

`of_seq_lwt` was mentioned above - that seems a little more clear than`inject` to me.

I will let you know when those changes are pushed.